meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch

```diff
@@ -0,0 +1,978 @@
++    bios_size = get_image_size(filename);
++    if (bios_size > 0) {
++        aligned_bios_size =
++            (bios_size + TARGET_PAGE_SIZE - 1) & TARGET_PAGE_MASK;
++
++        memory_region_init_ram(prom, NULL, "leon.bios", aligned_bios_size,
++                               &error_abort);
++        memory_region_set_readonly(prom, true);
++        memory_region_add_subregion(address_space_mem, 0x00000000, prom);
++
++        ret = load_image_targphys(filename, 0x00000000, bios_size);
++        if (ret < 0 || ret > bios_size) {
++            fprintf(stderr, "qemu: could not load prom '%s'\n", filename);
++            exit(1);
++        }
++    } else if (kernel_filename == NULL) {
++        fprintf(stderr, "Can't read bios image %s\n", filename);
++        exit(1);
++    }
++
++
++    memory_region_init_io(iomem, NULL, &leon_io_ops, s, "leon_io", 0x1000);
++    memory_region_add_subregion(get_system_memory(), 0x80000000, iomem);
++
++    leon_timer_init(&s->timer1, cpu_irqs[8], 1 /* id */);
++    leon_timer_init(&s->timer2, cpu_irqs[9], 2 /* id */);
++
++    if (serial_hd(0)) {
++        leon_uart_init(serial_hd(0), &s->uart1, cpu_irqs[3]);
++    }
++    if (serial_hd(1)) {
++        leon_uart_init(serial_hd(1), &s->uart2, cpu_irqs[2]);
++    }
++
++    /* HostFS */
++    hostfs_create(0x80001000, get_system_memory());
++
++    /* Initialize the GnatBus Master */
++    gnatbus_master_init(cpu_irqs, MAX_PILS);
++    gnatbus_device_init();
++
++    /* Can directly load an application. */
++    if (kernel_filename != NULL) {
++        long     kernel_size;
++        uint64_t entry;
++
++        kernel_size = load_elf(kernel_filename, NULL, NULL, NULL, &entry,
++                               NULL, NULL, 1 /* big endian */, EM_SPARC, 0,
++                               0);
++        if (kernel_size < 0) {
++            fprintf(stderr, "qemu: could not load kernel '%s'\n",
++                    kernel_filename);
++            exit(1);
++        }
++        if (bios_size <= 0) {
++            /* If there is no bios/monitor, start the application.  */
++            env->pc = entry;
++            env->npc = entry + 4;
++            reset_info->entry = entry;
++        }
++    }
++}
++
++static void at697_generic_machine_init(MachineClass *mc)
++{
++    mc->desc = "Leon-2 Atmel 697";
++    mc->init = at697_hw_init;
++    mc->default_cpu_type = SPARC_CPU_TYPE_NAME("LEON2");
++}
++
++DEFINE_MACHINE("at697", at697_generic_machine_init)
+diff --git a/hw/sparc/trace-events b/hw/sparc/trace-events
+index 355b07ae05..2856251f66 100644
+--- a/hw/sparc/trace-events
++++ b/hw/sparc/trace-events
+@@ -19,3 +19,16 @@ sun4m_iommu_bad_addr(uint64_t addr) "bad addr 0x%"PRIx64
+ # leon3.c
+ leon3_set_irq(int intno) "Set CPU IRQ %d"
+ leon3_reset_irq(int intno) "Reset CPU IRQ %d"
++
++# hw/leon.c
++leon_check_irqs(uint32_t pend, uint32_t force, uint32_t mask, uint32_t lvl1, uint32_t lvl2) "pend:0x%04x force:0x%04x mask:0x%04x lvl1:0x%04x lvl0:0x%04x\n"
```

Comment:
Maybe `\n` here is the issue?

Reply:
You mean, the very end of `\n` at `0x%04x\n"`?

Still trying to build it locally but haven't been able to successfully build it yet.  And it's taking ages...

Reply:
> You mean, the very end of `\n` at `0x%04x\n"`?

Yes

Reply:
Removed.